monitor/format: use `MonitorFormatter` to print on any `bufio.Writer` and not just on Stdout

Andreagit97 opened the pull request:
Running bpf tests (see here https://github.com/cilium/cilium/pull/39919), I realized that today we print only `DebugMsg` (see the last commit in this PR). This could be misleading when we receive a different event like `TraceNotify` because it will be erroneously decoded as a Debug message.



So I searched if today we have a method to decode all types and obtain a unique message in a string, so that we can use the output in `t.Log()`.

The `MonitorFormatter` provides a method `FormatSample` that does all the decoding but prints the final message to stdout. We would avoid this in tests, we would print to the console only in case of failure or with `-v`. 
So the idea was to provide the `MonitorFormatter` with `*bufio.Writer` so that we can choose at init time where to dump messages. 

In doing that, I decided to create a common interface for monitor's messages, hoping this could help future extensibility and reduce ad-hoc methods for different types of messages.

Let me know what you think about this approach :) Reviewing it commit by commit should be easier!

Andreagit97 committed:
monitor: create a common Dump method

The idea is to create a common Dump method among all the message types
so that it will be easier in the next commits to add an interface

Signed-off-by: Andrea Terzolo <andreaterzolo3@gmail.com>

Andreagit97 committed:
monitor: implement the `MonitorEvent` interface

Implement the `MonitorEvent` interface for all the monitor message types

Signed-off-by: Andrea Terzolo <andreaterzolo3@gmail.com>

Andreagit97 committed:
monitor:format: adopt the `MonitorEvent` interface

Signed-off-by: Andrea Terzolo <andreaterzolo3@gmail.com>

Andreagit97 committed:
monitor:hubble: remove redundant methods in favor of Decode

Signed-off-by: Andrea Terzolo <andreaterzolo3@gmail.com>

Andreagit97 committed:
monitor: use `MonitorFormatter` to log unknown events

Signed-off-by: Andrea Terzolo <andreaterzolo3@gmail.com>

Andreagit97 committed:
monitor: early return in `Dissect` method

Signed-off-by: Andrea Terzolo <andreaterzolo3@gmail.com>

Andreagit97 committed:
bpf/tests: use `MonitorFormatter` in bpf_test

Signed-off-by: Andrea Terzolo <andreaterzolo3@gmail.com>

kaworu commented on pkg/monitor/datapath_trace.go:
```diff
@@ -57,8 +56,31 @@ type TraceNotify struct {
+	}
+}
+
+// GetSrc retrieves the sorce endpoint for the message.
```
typo
```suggestion
// GetSrc retrieves the source endpoint for the message.
```

andreaterzolo replied:
Done, thanks!

kaworu commented on pkg/monitor/api/monitor_event_interface.go:
```diff
@@ -0,0 +1,84 @@
+type MonitorEvent interface {
+	// Decode decodes the message in 'data' into the struct.
+	Decode(data []byte) error
+	// GetSrc retrieves the sorce endpoint for the message
```
typo
```suggestion
	// GetSrc retrieves the source endpoint for the message
```

kaworu commented on pkg/monitor/api/monitor_event_interface.go:
```diff
@@ -0,0 +1,84 @@
+// DefaultSrcDstGetter is a default implementation of the GetSrc and GetDst methods
+type DefaultSrcDstGetter struct{}
+
+// GetSrc retrieves the sorce endpoint for the message
```
typo
```suggestion
// GetSrc retrieves the source endpoint for the message
```

kaworu commented on pkg/monitor/api/types.go:
```diff
@@ -193,6 +195,32 @@ type AgentNotify struct {
+	return dec.Decode(a)
+}
+
+// GetSrc retrieves the sorce endpoint for the message
```
typo
```suggestion
// GetSrc retrieves the source endpoint for the message
```

kaworu commented on pkg/monitor/datapath_debug.go:
```diff
@@ -239,12 +240,31 @@ type DebugMsg struct {
+	}
+}
+
+// GetSrc retrieves the sorce endpoint for the message.
```
typo
```suggestion
// GetSrc retrieves the source endpoint for the message.
```

kaworu commented on pkg/monitor/datapath_debug.go:
```diff
@@ -439,12 +444,31 @@ type DebugCapture struct {
+	}
+}
+
+// GetSrc retrieves the sorce endpoint for the message.
```
typo
```suggestion
// GetSrc retrieves the source endpoint for the message.
```

kaworu commented on pkg/monitor/datapath_drop.go:
```diff
@@ -63,22 +62,41 @@ type DropNotify struct {
+	}
+}
+
+// GetSrc retrieves the sorce endpoint for the message.
```
typo
```suggestion
// GetSrc retrieves the source endpoint for the message.
```

kaworu commented on pkg/monitor/datapath_policy.go:
```diff
@@ -62,12 +61,23 @@ type PolicyVerdictNotify struct {
 }
 
-func (n *PolicyVerdictNotify) decodePolicyVerdictNotify(data []byte) error {
+// GetSrc retrieves the sorce endpoint for the message.
```
typo
```suggestion
// GetSrc retrieves the source endpoint for the message.
```

kaworu commented on pkg/monitor/logrecord.go:
```diff
@@ -18,6 +22,32 @@ type LogRecordNotify struct {
+	}
+}
+
+// GetSrc retrieves the sorce endpoint for the message
```
typo
```suggestion
// GetSrc retrieves the source endpoint for the message
```

kaworu commented on pkg/monitor/api/types.go:
```diff
@@ -193,6 +195,32 @@ type AgentNotify struct {
+// GetDst retrieves the destination endpoint for the message.
+func (a *AgentNotify) GetDst() uint16 {
+	return 0
+}
```
Any reason to explicitly implement `GetSrc()` and `GetDst()` instead of embeding `DefaultSrcDstGetter` in `AgentNotify` if they always both return zero?

Is it because `(*AgentNotify).Decode` uses `gob.NewDecoder`? If correct, can you add a comment to explain this limitation please?

andreaterzolo replied:
Nope, it was just a leftover. I added the `DefaultSrcDstGetter` and wrote a test just to be sure the Decoding is still ok. ty!

kaworu commented on pkg/monitor/datapath_policy.go:
```diff
@@ -62,12 +61,23 @@ type PolicyVerdictNotify struct {
+
+// GetDst retrieves the destination endpoint for the message.
+func (n *PolicyVerdictNotify) GetDst() uint16 {
+	return uint16(n.RemoteLabel)
```
I believe `RemoteLabel` is not an endpoint identity, but a security identity. Also IIUC for `POLICY_INGRESS` it is the source label, and in that case the src/dst nomenclature doesn't apply cleanly (see [here](https://github.com/cilium/cilium/blob/1a6fbe49a84450cf528247d212796d335b8e2be8/bpf/bpf_lxc.c#L1990-L1993) for example).

andreaterzolo replied:
Thank you, I didn't notice that! I adapted the comment, let me know :)

kaworu commented on pkg/monitor/datapath_trace.go:
```diff
@@ -57,8 +56,31 @@ type TraceNotify struct {
+}
+
+// GetDst retrieves the destination endpoint for the message.
+func (tn *TraceNotify) GetDst() uint16 {
+	return tn.DstID
+}
```
AFAIK for `TRACE_TO_PROXY` traces `DstID` is the proxy port, see [here](https://github.com/cilium/cilium/blob/1a6fbe49a84450cf528247d212796d335b8e2be8/bpf/lib/trace.h#L14) and [here](https://github.com/cilium/cilium/blob/1a6fbe49a84450cf528247d212796d335b8e2be8/bpf/bpf_lxc.c#L671-L673).

andreaterzolo replied:
Thank you! I left a small comment on the method

kaworu commented on pkg/monitor/dissect.go:
might conflict with https://github.com/cilium/cilium/pull/37634 but this one will probably be merged first.

kaworu reviewed (approved):
Thanks for the PR @Andreagit97. Hubble bits LGTM. Some typos on the way and comments about the `GetSrc()`/`GetDst()` func. The confusion about "dst" is already there before the patch, so non-blocking.

kaworu reviewed (approved):
Thanks for the quick update @Andreagit97! Patch LGTM now, but the PR needs to be rebased as conflicts resolved before we can run CI.

smagnani96 commented on pkg/monitor/datapath_debug.go:
```diff
@@ -239,12 +240,31 @@ type DebugMsg struct {
+		// We don't print messages at INFO level
+		return
+	case api.JSON:
+		fmt.Fprintln(args.Buf, n.getJSON(args.CpuPrefix, args.LinkMonitor))
```
commit `monitor: allow formatting events on any bufio.Writer` has changes that could go in previous cleanup/refactor commit:

* `UnknownEvent/LostEvent` are renamed
* `DebugMsg` has some methods removed and being inlined (ex `DumpJSON`)

smagnani96 commented on pkg/monitor/datapath_debug.go:
```diff
@@ -239,12 +240,31 @@ type DebugMsg struct {
+	}
+}
+
+// GetSrc retrieves the source endpoint for the message.
```
In general, I'd advise to modify directly original commits with needed/suggested changes.

smagnani96 commented on pkg/monitor/datapath_debug.go:
```diff
@@ -239,12 +240,31 @@ type DebugMsg struct {
 
-func (n *DebugMsg) decodeDebugMsg(data []byte) error {
+// GetDst retrieves the destination endpoint for the message.
+func (n *DebugMsg) GetDst() uint16 {
```
Could reuse `api.DefaultSrcDstGetter` and then override only the `GetSrc` method.
Same applies to `DebugCapture`.

smagnani96 commented on pkg/monitor/dissect.go:
```diff
@@ -220,52 +214,52 @@ func GetConnectionSummary(data []byte, opts *decodeOpts) string {
-		defer dissectLock.Unlock()
-
-		initParser()
+func Dissect(buf *bufio.Writer, dissect bool, data []byte) {
```
Commit `cleanup: early return in Dissect method`: I also prefer early exists, but just to make sure this is only a refactor right? I don't see functional changes wrt the previous if-else.

We can keep the commit, would just rename it `monitor: early return in Dissect method` or `monitor:refactor: unwind if-else in Dissect method`.

andreaterzolo replied:
> I also prefer early exists, but just to make sure this is only a refactor right? 

Yep, this commit is just a refactor

smagnani96 commented on pkg/monitor/api/types_test.go:
```diff
@@ -4,6 +4,8 @@
 package api
```
I'd rename commit `cleanup: use MonitorFormatter to log unknown events` -> `monitor: use MonitorFormatter to log unknown events` so that it is more immediate to retrieve the unit it applies to.

smagnani96 replied:
Same `cleanup: remove redundant methods` -> `monitor:hubble: remove redundant methods in favor of Decode`

smagnani96 replied:
Same `format: adopt the MonitorEvent interface` `monitor:format: adopt the MonitorEvent interface`

smagnani96 reviewed (approved):
Patch LGTM, thanks Andrea!
I just left a couple of comments concerning commit naming and resolution of feedback/review (last commit).

gentoo-root commented on pkg/monitor/api/monitor_event_interface.go:
Please move the changes from this commit to the relevant commits where those changes are introduced. E.g., no need to make a typo and fix that typo in two different commits in a single PR.

andreaterzolo replied:
done :)

Andreagit97 committed:
monitor: allow formatting events on any `bufio.Writer`

Actually we support only stdout formatting, this is not handy in some
scenarios. For example if we want to enable logs in tests using `t.Log`
or if we want to assert against the content of a message. This commit
introduces the possibility to dump messages on any `bufio.Writer`

Signed-off-by: Andrea Terzolo <andreaterzolo3@gmail.com>
